src/NonCanonicalReader.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace PhpSchool\Terminal;
+
+/**
+ * This class takes a terminal and disabled canonical mode. It reads the input
+ * and returns characters and control sequences as `InputCharacters` as soon
+ * as they are read - character by character.
+ *
+ * On destruct canonical mode will be enabled if it was when in it was constructed.
+ *
+ * @author Aydin Hassan <aydin@hotmail.co.uk>
+ */
+class NonCanonicalReader
```

Comment:
So this new class takes a `Terminal` object and allows to read input and maps to `InputCharacters` - it also allows to map other characters to control code. So `cli-menu` code will look like this:

```php
$reader = new NonCanonicalReader($this->terminal);
$reader->addControlMappings([
    '^P' => InputCharacter::UP,
    'k'  => InputCharacter::UP,
    '^K' => InputCharacter::DOWN,
    'j'  => InputCharacter::DOWN,
    "\r" => InputCharacter::ENTER,
    ' '  => InputCharacter::ENTER,
]);

while ($this->isOpen() && $char = $reader->readCharacter()) {
    if ($char->isNotControl()) {
        continue;
    }

    switch ($char->getControl()) {
        case InputCharacter::UP:
        case InputCharacter::DOWN:
            $this->moveSelection($char->getControl());
            $this->draw();
            break;
        case InputCharacter::ENTER:
            $this->executeCurrentItem();
            break;
    }
}
```
It enables and disables canonical mode on the fly and empties the buffer on each read (well reads 4 bytes but there should never be more in there)

What do you think?